File: javaLeetcode/binaryTreeRelated.java
package javaLeetcode;

import java.util.*;

/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public List<Integer> preorderTraversal(TreeNode root) {
        if (root == null) {
            return new ArrayList<>();
        }
        List<Integer> ret = new ArrayList<>();
        ret.add(root.val);
        ret.addAll(preorderTraversal(root.left));
        ret.addAll(preorderTraversal(root.right));
        return ret;
    }

    public List<Integer> postorderTraversal(TreeNode root) {
        if (root == null) {
            return new ArrayList<>();
        }
        List<Integer> ret = new ArrayList<>();
        ret.addAll(postorderTraversal(root.left));
        ret.addAll(postorderTraversal(root.right));
        ret.add(root.val);
        return ret;
    }

    public boolean isBalanced(TreeNode root) {
        if (root == null) 
            return true;
        if ( height(root) == -1) 
            return false;
        return true;
    }

    int height(TreeNode node) {
       if (node == null) {
        return 0;
       } 
       int leftHeight = height(node.left); 
       int rightHeight = height(node.right); 

       if (leftHeight == -1 || rightHeight == -1)
        return -1;
       if (Math.abs(leftHeight- rightHeight) > 1) 
        return -1;
       return Math.max(leftHeight, rightHeight) + 1;
    }

    public TreeNode invertTree(TreeNode root) {
        TreeNode tmp = root.left;
        root.left = root.right;
        root.right = tmp;
        invertTree(root.left);
        invertTree(root.right);
        return root;
    }

    public List<String> binaryTreePaths(TreeNode root) {
        List<String> paths = new ArrayList<>();
        if (root == null) {
            return paths;
        }
        if (root.left == null && root.right == null) {
            paths.add(String.valueOf(root.val));
            return paths;
        }
        List<String> leftPath = binaryTreePaths(root.left);
        List<String> rightPath = binaryTreePaths(root.right);
        for (String path:leftPath) {
            paths.add(String.valueOf(root.val) + "->" + path);
        }
        for (String path:rightPath) {
            paths.add(String.valueOf(root.val) + "->" + path);
        }
        return paths;
    }
}

class TreeNode {
    int val;
    TreeNode left;
    TreeNode right;
    TreeNode() {}
    TreeNode(int val) { this.val = val; }
    TreeNode(int val, TreeNode left, TreeNode right) {
        this.val = val;
        this.left = left;
        this.right = right;
    }
}

File: go/binaryTreeRelated.go
package main

import (
	"math"
)

func findTilt(root *TreeNode) int {
	if root == nil {
		return 0
	}
	return abs(findSum(root.Left)-findSum(root.Right)) + findTilt(root.Left) + findTilt(root.Right)
}

func findSum(node *TreeNode) int {
	if node == nil {
		return 0
	}
	return node.Val + findSum(node.Left) + findSum(node.Right)
}

func abs(num int) int {
	if num < 0 {
		return 0 - num
	}
	return num
}

// Definition for a binary tree node.
type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func constructMaximumBinaryTree(nums []int) *TreeNode {
	if len(nums) == 0 {
		return nil
	}
	index, max := findLargest(nums)
	node := TreeNode{
		Val:   max,
		Left:  nil,
		Right: nil,
	}
	if len(nums) == 1 {
		return &node
	}
	node.Left = constructMaximumBinaryTree(nums[:index])
	node.Right = constructMaximumBinaryTree(nums[index+1:])
	return &node
}

func findLargest(nums []int) (index int, max int) {
	for i := 0; i < len(nums); i++ {
		if nums[i] > max {
			max = nums[i]
			index = i
		}
	}
	return index, max
}

func printTree(root *TreeNode) [][]string {
	height := getHeight(root)
	m := height + 1
	n := int(math.Exp2(float64(height+1))) - 1
	ret := make([][]string, m)
	for i := range ret {
		ret[i] = make([]string, n)
	}
	placeToMatrix(root, ret, m, (n-1)/2, 0)
	return ret
}

func placeToMatrix(node *TreeNode, matrix [][]string, height int, c int, row int) {
	if node == nil {
		return
	}
	if row >= height {
		return
	}
	if row == 0 {
		matrix[row][c] = string(node.Val)
	}
	matrix[row+1][c-int(math.Exp2(float64(c-row-1)))] = string(node.Left.Val)
	matrix[row+1][c+int(math.Exp2(float64(c-row-1)))] = string(node.Right.Val)
	placeToMatrix(node.Left, matrix, height, c+int(math.Exp2(float64(c-row-1))), row+1)
	placeToMatrix(node.Right, matrix, height, c-int(math.Exp2(float64(c-row-1))), row+1)
}

func getHeight(root *TreeNode) int {
	if root == nil {
		return 0
	}
	if getHeight(root.Left) > getHeight(root.Right) {
		return 1 + getHeight(root.Left)
	}
	return 1 + getHeight(root.Right)
}
